Update go codegen

saicheems opened the pull request:
Fixes some build errors, removes the cloud package from the public
surface.
Also changed the snippet file to only use spaces. It makes the editing
process a lot easier, and `gofmt` can convert to tabs through a cleanup
step in the pipeline.

saicheems commented:
@jmuk 
@jba 
@pongad 
@vchudnov-g

saicheems commented on vgen/src/main/resources/io/gapi/vgen/go/main.snip:
```diff
@@ -210,8 +210,8 @@
-                    	callOpts := append({@retryCodesMethodName}RetryCodes(), {@retryParamsMethodName}RetryParams(), client.CallOptions["{@methodName}"]...)
-                    	callOpts = append(callOpts, opts...)
+                    	defaultCallOpts := append({@retryCodesMethodName}RetryCodes(), {@retryParamsMethodName}RetryParams()...)
+                    	callOpts := append(client.CallOptions["{@methodName}"], opts...)
```
Not super satisfied with all the appends thrown around... but I couldn't think of a better way (other than making a function that takes an arbitrary number of `[]CallOption`).

LMK

pongad commented on vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline:
```diff
@@ -191,14 +189,14 @@ func (book Book) Path() (string, error) {
-	callOpts := append(nonIdempotentRetryCodes(), defaultRetryParams(), client.CallOptions["CreateShelf"]...)
-	callOpts = append(callOpts, opts...)
+	defaultCallOpts := append(nonIdempotentRetryCodes(), defaultRetryParams()...)
+	callOpts := append(client.CallOptions["CreateShelf"], opts...)
```
I am a little worried about this. If `client.CallOptions[s]` is not a filled slice, we will have a race condition if two goroutines call the same method on the same client concurrently. See example in https://play.golang.org/p/tgObgdGEsG .

It probably also worth documenting that the user may not assign/modify client.CallOptions concurrently with any calls on the client.

saicheems replied:
Wow yeah nice catch!

On inspection I think the right course of action is to not export `client.CallOptions` given that it really shouldn't be modifiable outside of initializing a new client struct.

There's still some work to be done on client configuration before it reaches parity with the python implementation... for now I'll make that struct var private.

pongad commented:
There's a potential surprising behavior I'm worried about, other than that I think it looks good

jmuk commented on vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline:
```diff
@@ -109,24 +107,24 @@ type Client struct {
 		conn: conn,
 		client: google_example_library_v1.NewLibraryServiceClient(conn),
-		CallOptions: map[string][]gax.CallOption{},
+		callOptions: map[string][]gax.CallOption{},
```
This callOptions is now hidden and always empty... Why does this exist? Is there a way this can be updated?
You can simply remove it if you don't need it anymore.
(or, probably add a new gax.ClientOption to reset the default callOptions?)

saicheems replied:
My plan is to allow it to be instantiated through GAX (WithRetryOptionsOverride)
Was also considering populating it to supply the defaults...

I'll leave it for now

jmuk replied:
I see, then that's good.

Please address my comment below about tricky behavior of append(). It actually depends on how callOptions will be constructed from the retry-options-override though.

saicheems replied:
Yup, will do

jmuk commented on vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline:
```diff
@@ -191,14 +189,14 @@ func (book Book) Path() (string, error) {
-	callOpts := append(nonIdempotentRetryCodes(), defaultRetryParams(), client.CallOptions["CreateShelf"]...)
-	callOpts = append(callOpts, opts...)
+	defaultCallOpts := append(nonIdempotentRetryCodes(), defaultRetryParams()...)
+	callOpts := append(client.callOptions["CreateShelf"], opts...)
```
This reminds me of a discussion in the internal code review. append() has an optimization that if the latter part (client.callOptions["CreateShelf"] in this case) has enough space, append() actually updates it rather than creating a new slice, thus this operation is not very safe sometimes. You may want to make a copy instead.

jba replied:
Yes, this is buggy. append does try to modify its first argument if there is space.

I suggest simple, linear code:

var callOpts []gax.CallOption
callOpts = append(callOpts, thing1)
callOpts = append(callOpts, thing2)
...

(This is not optimal in space or time, but it almost certainly won't matter.)

saicheems replied:
Makes sense, fixed.

saicheems commented:
Introduced a lot of noise just now - I converted all tabs in the snippet file to spaces. I ran into a bug due to mix and matching... When `gofmt` is integrated into the pipeline, it will not be an issue.

All other concerns are addressed.

jmuk commented:
LGTM. I'm a bit concerned about using spaces, but actually artman runs gofmt, so you don't have to worry about this much (see https://github.com/googleapis/artman/blob/master/pipeline/pipelines/code_generation_pipeline.py#L274)

pongad commented on vgen/src/test/java/io/gapi/vgen/testdata/go_library.baseline:
```diff
@@ -52,226 +52,236 @@ import (
+    callOpts = append(callOpts, nonIdempotentRetryCodes())
+    callOpts = append(callOpts, defaultRetryParams())
+    callOpts = append(callOpts, client.callOptions["CreateShelf"])
+    callOpts = append(callOpts, opts...)
```
I just realized we can use gax.callOptions if it were exported. I'm not sure if it's worth changing at this point.

saicheems replied:
I'll leave it private in this CL, my plan right now is to make `ClientSettings`/`CallSettings` public in `gax`, and set `client.callOptions` during initialization.

There shouldn't really be a pathway imo for users to modify client level settings like retry overrides (the intended use case of `client.callOptions` outside of initialization.

saicheems commented:
Yeah the snippet language leaves us in an odd state where using all tabs is buggy, and a mix of tabs and spaces is annoying to develop with and also buggy...
I think it should be fine if artman runs `gofmt` (cool!)
